hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/DispatchingEvmFrameState.java

```diff
@@ -87,72 +87,90 @@ public class DispatchingEvmFrameState implements EvmFrameState {
     private static final String TOKEN_CALL_REDIRECT_CONTRACT_BINARY =
             "6080604052348015600f57600080fd5b506000610167905077618dc65efefefefefefefefefefefefefefefefefefefefe600052366000602037600080366018016008845af43d806000803e8160008114605857816000f35b816000fdfea2646970667358221220d8378feed472ba49a0005514ef7087017f707b45fb9bf56bb81bb93ff19a238b64736f6c634300080b0033";
 
-    private final Dispatch dispatch;
-    private final WritableKVState<SlotKey, SlotValue> storage;
-    private final WritableKVState<EntityNumber, Bytecode> bytecode;
+    private final HederaNativeOperations nativeOperations;
+    private final ContractStateStore contractStateStore;
 
     public DispatchingEvmFrameState(
-            @NonNull final Dispatch dispatch,
-            @NonNull final WritableKVState<SlotKey, SlotValue> storage,
-            @NonNull final WritableKVState<EntityNumber, Bytecode> bytecode) {
-        this.storage = requireNonNull(storage);
-        this.bytecode = requireNonNull(bytecode);
-        this.dispatch = requireNonNull(dispatch);
+            @NonNull final HederaNativeOperations nativeOperations,
+            @NonNull final ContractStateStore contractStateStore) {
+        this.nativeOperations = requireNonNull(nativeOperations);
+        this.contractStateStore = requireNonNull(contractStateStore);
     }
 
     @Override
     public void setStorageValue(final long number, @NonNull final UInt256 key, @NonNull final UInt256 value) {
         final var slotKey = new SlotKey(number, tuweniToPbjBytes(requireNonNull(key)));
-        final var oldSlotValue = storage.get(slotKey);
-        // Ensure we don't change any prev/next keys until committing the final WorldUpdater
+        final var oldSlotValue = contractStateStore.getSlotValue(slotKey);
+        // Ensure we don't change any prev/next keys until the base commit
         final var slotValue = new SlotValue(
                 tuweniToPbjBytes(requireNonNull(value)),
                 oldSlotValue == null ? com.hedera.pbj.runtime.io.buffer.Bytes.EMPTY : oldSlotValue.previousKey(),
                 oldSlotValue == null ? com.hedera.pbj.runtime.io.buffer.Bytes.EMPTY : oldSlotValue.nextKey());
-        storage.put(slotKey, slotValue);
+        // We don't call remove() here when the new value is zero, again because we
+        // want to preserve the prev/next key information until the base commit; only
+        // then will we remove the zeroed out slot from the K/V state
+        contractStateStore.putSlot(slotKey, slotValue);
```

Comment:
Ah.  This answers that question then about `removeSlot`|`putSlot` above: `putSlot` puts zeros.  But makes the question of `getNumSlots` more interesting.